pkg/ingester/ingester_v2.go

```diff
@@ -452,3 +461,63 @@ func (i *Ingester) closeAllTSDB() {
 	i.userStatesMtx.Unlock()
 	wg.Wait()
 }
+
+// openExistingTSDB walks the user tsdb dir, and opens a tsdb for each user. This may start a WAL replay, so we limit the number of
+// concurrently opening TSDB.
+func (i *Ingester) openExistingTSDB(ctx context.Context) error {
+	wg := &sync.WaitGroup{}
+	openGate := gate.New(i.cfg.TSDBConfig.MaxOpeningTSDBOnStartup)
+
+	err := filepath.Walk(i.cfg.TSDBConfig.Dir, func(path string, info os.FileInfo, err error) error {
+		if path == i.cfg.TSDBConfig.Dir { // Nothing to do for root
+			return nil
+		}
+
+		// Top level directories are assumed to be user TSDBs
+		if info.IsDir() {
+
+			userID := info.Name()
+			f, err := os.Open(path)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "unable to open user TSDB dir", "err", err, "user", userID)
+				return filepath.SkipDir
+			}
+			defer f.Close()
+
+			// If the dir is empty skip it
+			if _, err := f.Readdirnames(1); err != nil {
+				if err != io.EOF {
+					level.Error(util.Logger).Log("msg", "unable to read TSDB dir", "err", err, "user", userID)
+					return filepath.SkipDir
+				}
+
+				// Empty dir
+				return filepath.SkipDir
+			}
+
+			// Limit the number of TSDB's opening concurrently
+			if err := openGate.Start(ctx); err != nil {
+				return err
+			}
+
+			wg.Add(1)
+			go func(userID string) {
+				defer wg.Done()
+				defer openGate.Done()
+				_, err := i.getOrCreateTSDB(userID, true) // force create the TSDB due to the lifecycler not having started yet
```

Comment:
I'm a bit confused. Given `getOrCreateTSDB()` wraps `tsdb.Open()` within the `userStatesMtx.Lock()`, and the WAL reply is done by `tsdb.Open()` before returning, how does the parallelization works and how did you hit the issue described in the PR anyway?

Reply:
Huh I totally missed that. The behavior I observed pointed to WAL replay eating memory, but you're right we can't be opening more than 1 at a time already. Guess I nees to try and recreate and get a memory profile.

Reply:
> get a memory profile

That would be very helpful. You can download an heap profile from the ingester's HTTP endpoint `/debug/pprof/`.

> I totally missed that

Please don't trash this PR. In any case I believe that pre-loading all TSDBs before an ingester is active is the desired behaviour, but right now it's not really parallelized.

Reply:
> In any case I believe that pre-loading all TSDBs before an ingester is active is the desired behaviour, but right now it's not really parallelized.

If end up to be complicated, we can also come up removing the concurrency option and pre-loading all TSDBs sequentially (which to me is better than now) and then add concurrency in a subsequent PR.

Reply:
I refactored `getOrCreateTSDB` so that we can call `createTSDB` in the go routines and then only lock to add it to the db list once open.